presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Pattern.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.base.MoreObjects;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class Pattern
+{
+    private Pattern() {}
+
+    public abstract boolean matches(PlanNode node);
+
+    public static Pattern any()
+    {
+        return AnyNode.INSTANCE;
+    }
+
+    public static <T extends PlanNode> Pattern node(Class<T> nodeClass)
+    {
+        return new ClassNodeBased(nodeClass);
+    }
+
+    static class AnyNode
+            extends Pattern
+    {
+        final static AnyNode INSTANCE = new AnyNode();
+
+        @Override
+        public boolean matches(PlanNode node)
+        {
+            return true;
+        }
+    }
+
+    static class ClassNodeBased<T extends PlanNode>
+            extends Pattern
+    {
+        private final Class<T> nodeClass;
+
+        ClassNodeBased(Class<T> nodeClass)
+        {
+            this.nodeClass = requireNonNull(nodeClass, "nodeClass is null");
+        }
+
+        Class<T> getNodeClass()
```

Comment:
Maybe there is no point in having `Any` pattern as separate instance. I think the problem you solve is very specific (e.g: do initial pre-matching on node class), so we probably don't want any other implementation of `Pattern` for this simple solution. E.g: maybe we shouldn't generalize very specific solution for now.